### ISSUE TITLE
Unsafe tool

### ISSUE THREAD

**Issue by nicolas-f** [Marginal/EDMarketConnector#89]
Hi,

I found that the elite dangerous user, password are stored in plain-text into the windows registry. You should really consider not storing it, or using a user master password in order to encrypt theses credentials.

**Comment by Marginal** [comment]
If you're running malware that has access to your user account then you're screwed and a master password isn't going to help. Instead, it's time to contact your bank and credit card companies and change all of your important online account passwords (probably including your Frontier password, 'though an attacker can do practically nothing of interest with it).

Alternatively, if the threat that you're concerned about isn't malware, then what is it?

**Comment by nicolas-f** [comment]
Ok, I understand that maybe this feature request is exaggerated. However, at least, the user should be warned that the password is stored in plain-text.

**Comment by Marginal** [comment]
Fine, I'll use the same text that IE or Chrome uses to warn the user that their passwords are trivially available to malware that has access to their user account.

Oh wait, there isn't any such warning.

**Comment by Scyntrus** [comment]
This really isn't acceptable practice.

With Chrome, I can choose not to save my password. I'm completely aware at how insecure it is and therefore don't save passwords. As well, the new version of Chrome integrates with Windows credential store and requires your Windows password before showing saved passwords.

I recommend adding an option to not save passwords AND using a library such as this to secure passwords: https://pypi.python.org/pypi/keyring/5.3

**Comment by Marginal** [comment]
> the new version of Chrome integrates with Windows credential store and requires your Windows password before showing saved passwords

Which [protects](https://code.google.com/p/chromium/issues/detail?id=353890#c5) you against a naive co-worker seeing your passwords while you're away from your desk having left your computer unlocked, but does nothing against malware (which can just call [CryptUnprotectData](https://code.google.com/p/chromium/codesearch#chromium/src/components/os_crypt/os_crypt_win.cc) without prompting) or against a co-worker who knows about chrome://flags/#password-manager-reauthentication. So is largely pointless.

> I recommend adding an option to not save passwords 

Patches welcome, I guess.

> AND using a library such as this to secure passwords: https://pypi.python.org/pypi/keyring/5.3

Also largely pointless because malware can just call [CredRead](https://github.com/jaraco/keyring/blob/master/keyring/backends/Windows.py#L65) to obtain the password.

**Comment by thommcgrath** [comment]
Don't let perfect be the enemy of good.

You are 100% correct that that a compromised computer is game over. You are correct that malware can break it pretty easily, at least on Windows. That doesn't mean you should take such a lax stance on the issue.

Using "secure" storage would be better. Not perfect, but still better.

**Comment by danmac-uk** [comment]
After seeing things blow up about this issue [over on Reddit](https://www.reddit.com/r/EliteDangerous/comments/5q8xkq/ed_market_connector_stores_login_credentials/), I'd be willing to throw a few hours at this. Would you be interested in a pull request that changes the current password storage to use the [keyring](https://pypi.python.org/pypi/keyring) package from PyPi?

I've skimmed the code and I'm inclined to bastardise [this bit here](https://github.com/Marginal/EDMarketConnector/blob/master/config.py#L218) so if the getter / setter is called with the key "password" then it uses the functionality provided in the keyring package for storage rather than the registry. This hopefully will solve the plaintext password issue without breaking anything else. I imagine a similar technique could be used for the Mac related code too. Not sure about Linux, I imagine there would need to be a bit more work to fallback gracefully in case KWallet / secretservice isn't installed. I can test on all three platforms, even the one nobody will be using ;) Of course you know your code better than I, so I'm happy to discuss / round down the fine details etc.

I may not get a chance to work on this immediately, and I'd also suggest providing the option to not save the password at all (which I haven't even looked at how to implement, nor am I familiar with the GUI toolkit) but at this stage I'm just wondering if you're interested in any help on this one?

FWIW, I'm happy to run the existing version as-is. If someone gets access to my machine on the level required to pull apart my user registry I have bigger problems. Having said that, there is room for improvement in the code here.

Cheers :)

**Comment by neotron** [comment]
A compromised account is compromised, but if, say, someone steals my computer and somehow managed to reset my password (it's doable using iCloud on Mac in some cases), they can log in as my account.

Using an unsecured password, they can access it freely but the keychain remains inaccessible to them due to it being encrypted with my login password. Windows, presumably, works the same.

Likewise using the keychain that's bound by login prevents another user on the same access to your credentials, i.e https://www.adminarsenal.com/admin-arsenal-blog/modifying-the-registry-of-another-user/ on Windows. Leaving it plaintext in the registry means that any user on the computer has free access.

Also as noted the better solution if you don't want to provide any security is to not save the password on disk at all. At least then it's just in-memory, not perfect but better than nothing.

**Comment by Marginal** [comment]
> Would you be interested in a pull request that changes the current password storage to use the keyring package from PyPi?

Yes. But against the multi_account branch.

>  I'd also suggest providing the option to not save the password at all

As I said [above](https://github.com/Marginal/EDMarketConnector/issues/89#issuecomment-183906612), patches welcome. Probably the simplest UI would be to allow the user to leave the password field blank in settings and, if unset, prompt for the password during AppWindow.login().

> Using an unsecured password, they can access it freely but the keychain remains inaccessible to them due to it being encrypted with my login password. Windows, presumably, works the same.

I can see an argument for using OSX's keychain, since it is reasonably secure against attack. Windows Credentials store is readable by any process running as the current user or with elevated privileges so adds no security at all - it's like putting all your eggs in a crappy basket.

**Comment by danmac-uk** [comment]
No worries bud, if you're happy with the general idea and how I initially envisage the bastardisation of your code I'll see if I can throw some time at this in the next week or two. I don't have a Windows Python (or Git, for that matter) environment setup, nor am I familiar with the GUI toolkit you're using (I used PySide a while ago, that's it) but the stuff you put in the README should be enough to get me going.

I agree with your assessment of the Windows credential store but I do feel it would be marginally better than plaintext in the registry. I'm not immediately aware of any better ideas. At least this should deter the most casual attacks.

Thanks for getting back to me so quick.

**Comment by Marginal** [comment]
> nor am I familiar with the GUI toolkit you're using

It's quite arcane, but you can copy how the companion.VerificationRequired exception is handled.

**Comment by danmac-uk** [comment]
OK I think I'm set up. Kind of. Cloned the repo, cloned it again with -b multi_account :P I tried to make this thing build an EXE and I'm pretty sure I need some Windows dev SDK / VS2008. EDMC itself runs fine via Python directly tho, so that'll do for now.

I got a "merge module" it needed from [here](https://github.com/OpenNI/OpenNI2/blob/master/Packaging/Install/Resources/Microsoft_VC90_CRT_x86.msm) which is probably of dubious origin and legality. But then it shat itself looking for some VBScripts in "C:\Program Files (x86)\Windows Kits\8.1\bin\x86".

BTW there is a newer WiX Toolset out, which I ended up unwittingly downloading. Seems to work OK. Changed setup.py to suit the different pathname but then it fell over on the above.

Thanks for the additional pointers BTW :) I see there are some relevant changes in the multi_account branch (predictably enough) it looks like you're using two lists now (users & passes) rather than single strings but I'll try to get up to speed with that in the next week or so, and have a tinker with the keyring package. If I can pull my finger out in time for the 2.26 release I might be able to hitch a ride on the migrate code you already did in prefs.py. Would be nice to have an uninstall hook to blow away the reg keys and keyring stuff but I'm hesitant to bite off more than I can chew already.

Anyway I had enough of Windows for one day, if I have to use that terminal for another minute I might have to punch myself in the eye :P Cheers.

**Comment by Marginal** [comment]
@danmac-uk 
keyring part is done. Are you wanting to do the "blank password" = prompt part?

Don't worry about not being able to package - it's not required for testing. Though it sounds like I need to add stuff about the merge modules to the packaging instructions.

**Comment by danmac-uk** [comment]
ahhh mate I was half way through that already :( although to be fair your way was neater than mine, I was keeping the original config calls and shuffling things behind the scenes.

I was less hesitant about the keyring stuff than I am about the GUI related stuff but I can have a look at it next week if I don't see any movement.

Some code here if you're interested, may be helpful if you get to it first.

Here is a bit of code that makes sure the keyring is working as expected (the idea was to disable save password option on systems that don't play ball):

(from line 172 of config.py at least on mine)
```
    elif platform=='win32':

        def __init__(self):

            # testing functionality of keyring
            self.can_save_password = False
            try:
                if isinstance(keyring.get_keyring(), keyring.backends.Windows.WinVaultKeyring):  # is the backend what we expect?
                    self.can_save_password = True
                else:
                    print("Keyring Error: WinVaultKeyring not available.")
            except Exception as e:
                print("Keyring Error: " + str(e))

            if self.can_save_password:
                try:
                    keyring.set_password("edmc","TeStInGpAsSwOrDvAuLt","z3r0c00l")
                    if keyring.get_password("edmc","TeStInGpAsSwOrDvAuLt") != "z3r0c00l":
                        print("Keyring Error: Stored password was munged.")
                        self.can_save_password = False
                    keyring.delete_password("edmc","TeStInGpAsSwOrDvAuLt")
                except Exception as e:
                    print("Keyring Error: " + str(e))
                    self.can_save_password = False

            if self.can_save_password:
                print("Keyring support enabled.")
            else:
                print("Keyring support disabled.")
```

Same code with different backend name should work for Mac but I hadn't yet set up a virtualenv etc on Mac. I was planning to move it out of there, I only stuck it in the win32 section cos that's what I was testing on.

Also add to bottom of README acknowledgements:

`* Uses [Python Keyring Lib](https://github.com/jaraco/keyring) by Jason R. Coombs, Kang Zhang, et al.
`

(there are some 2016's I updated to 2017 also in the README)

Oh yeah and of course add to requirements:

`keyring >= 10.2, < 11.0
`

Anyway yeah nice job cheers, have a good weekend.

**Comment by Marginal** [comment]
You obviously code more defensively than me - I tend to assume that if an API returns success then it was successful.

The code warns about insecure backends [here](https://github.com/Marginal/EDMarketConnector/blob/multi_account/EDMarketConnector.py#L296) but I don't expect users to see this on Windows or macOS since the backends on those platforms are talking to system services.

The code *does* assume that users don't deny access to the keychain on platforms where they have that choice (macOS, Linux) and the app will quit or give an error if they don't. That's something that will need to addressed if users can leave the password setting blank in order to be prompted for it as needed.

I've added the acknowledgement in 2952fc0, thanks for the reminder.

**Comment by danmac-uk** [comment]
Haha not always :P But in situations like this (security-related, not my codebase, not experienced with the keyring package, run by all sorts of randoms on all sorts of wacky platforms, etc) I like to establish a nice warm safety blanket. Particularly where security is concerned I always have an eye on ensuring things I expect to happen definitely happen the way I expected.

I doubt those simple init tests would be useful in 99% of circumstances (although a simple boolean "does the keyring work" I imagine will be useful down the road, particularly cross-platform) but I'll give you a for-instance. Well over a decade ago now, I wrote a similarly paranoid check into some code to avoid a theoretical scenario I never expected to actually happen, but if it did it would have been less than ideal. May have been a race condition, I forget. I used an error message I knew would get a reaction. Imagine my surprise when I had a rather confused phone call from a user asking if there really had been an unexpected fault in the space-time continuum and if it was really necessary to verify his grandfather is who he remembers him to be ... :P Only happened the once, though.
